### PR TITLE
Show shared forms on navigation

### PR DESCRIPTION
**Pull request by jotoeri**
Show the shared forms on navigation. Implements #364

- [x] API Docs are still missing.
- [x] Probably better split API to have getForms & getSharedForms, which might somewhen be extended with e.g. shared Results

grafik
grafik

### CODE REVIEW

**Comment by jotoeri** [comment]
Hm, @jancborchardt are you sure, you want it to have like this? I could imagine this becomes a bit confusing if the list becomes longer. Especially if we somewhen add further lists like for Result-Sharing or a collapsed one for kind of an archive?
The old way it was collapsable from the vue-components. Wouldn't it be better if we just use the component and probably reduce the inset/change the style there a bit (effecting of course all using apps)?
grafik

**Comment by jotoeri** [comment]
And for reference the same vue-components change (half inset) also on deck and polls. 😉 I really like it like this!
grafik
grafik
grafik

**Comment by skjnldsv** [comment]
Should probably be "My forms" instead. cc @jancborchardt

**Comment by jotoeri** [comment]
> 
> 
> Should probably be "My forms" instead. cc @jancborchardt

Yeah, if you have a look onto the first two pictures, this is already 'Your Forms' and 'Shared with you'.

I think the most important question is for now: How do we list the forms on navigation? @jancborchardt @skjnldsv 
- Collapsable Navigation Item? Maybe with a decreased inset, as on the last picture of my previous comment? @skjnldsv Do you think we would get that through on vue-components? (This is my favourite)
- Non-Collapsable just listed with AppNavigationCaption (w/o icon) or with AppNavigationItem (w/ icon)

**Comment by jancborchardt** [comment]
I think this one you posted looks most proper @jotoeri? It’s similar to the Settings with "Personal" and "Admin" headings in the nav, isn’t that in the components as well? Cause the headings themselves actually do not need to be _selectable_ entries, just separators.

> image

**Comment by jotoeri** [comment]
Yes, it's also a component. I would have liked more the collapsable solution. But ok. Now its the Caption-Version.

grafik
grafik

**Comment by skjnldsv** [comment]
If "your forms" is empty, we should not display an empty list, I think we should add a small message?

**Comment by jotoeri** [comment]
Yeah, had that also in mind... But would need probably some AppNavigationEmptyContent on components?! Seems like the AppNavigation only takes appropriate components inside the template, no?

**Comment by skjnldsv** [comment]
> But would need probably some AppNavigationEmptyContent on components?!

I don't think we'd have too much usage for it. Maybe just remove the title and spacing? Just show th shared ones?

The main content should already state that you don't have any and you can create new ones, no?

**Comment by jotoeri** [comment]
Yeah, you're right. The Spacer i just removed completely as i realised, the Caption does that itself. So no need to have twice a space.
grafik
grafik

File: appinfo/routes.php [hunk 1]
```diff
@@ -25,19 +25,16 @@
...
-		['name' => 'page#index', 'url' => '/new', 'verb' => 'GET', 'postfix' => 'create'],
-		['name' => 'page#index', 'url' => '/{hash}/edit', 'verb' => 'GET', 'postfix' => 'edit'],
-		['name' => 'page#index', 'url' => '/{hash}/clone', 'verb' => 'GET', 'postfix' => 'clone'],
-		['name' => 'page#index', 'url' => '/{hash}/results', 'verb' => 'GET', 'postfix' => 'results'],
-
+		// Before /{hash}/{action} to avoid conflict
 		['name' => 'page#goto_form', 'url' => '/{hash}', 'verb' => 'GET'],
+
+		// As parameters have defaults, this catches all routes from '/' to '/hash/edit'
+		['name' => 'page#index', 'url' => '/{hash}/{action}', 'verb' => 'GET', 'defaults' => ['hash' => '', 'action' => '']],
```
Comment (skjnldsv) [inline appinfo/routes.php:32]: nice!

File: src/Forms.vue [hunk 2]
```diff
@@ -203,10 +246,10 @@ export default {
 		 * @param {Number} id the form id
 		 */
 		async onDeleteForm(id) {
-			const formIndex = this.forms.findIndex(form => form.id === id)
-			const deletedHash = this.forms[formIndex].hash
+			const formIndex = this.forms.owned.findIndex(form => form.id === id)
+			const deletedHash = this.forms.owned[formIndex].hash
 
-			this.forms.splice(formIndex, 1)
+			this.forms.owned.splice(formIndex, 1)
```
Comment (skjnldsv) [inline src/Forms.vue:252]: I think I would have been more confident having two separate array for this, `forms` and `sharedForms`

File: src/Forms.vue [hunk 3]
```diff
@@ -106,27 +120,45 @@ export default {
 		return {
 			loading: true,
 			sidebarOpened: true,
-			forms: [],
+			forms: {
+				owned: [],
+				shared: [],
+			},
```
Comment (skjnldsv) [inline src/Forms.vue:126]: I think I would have been more confident having two separate array for this, `forms` and `sharedForms`

I'm really not fond of nested values in data.

**Review by skjnldsv — CHANGES_REQUESTED** [review]

File: src/Forms.vue (hunk 3 above)
Comment (jotoeri) [reply src/Forms.vue]: 🤷‍♂️ Changed that...

**Comment by jotoeri** [comment]
Just thougth about the access stuff again, to fix the redirect-thing #539. As it also clarifies/secures the work & code here, i squashed it in here, instead of a separate PR, that would change the changes of this PR afterwards.
Sorry for this change after the first review. The result here just seems to be more clear & clean to me. 🙈

File: src/Forms.vue [hunk 4]
```diff
@@ -26,17 +26,28 @@
...
 					@clone="onCloneForm"
 					@delete="onDeleteForm" />
+
+				<!-- Shared Forms-->
+				<AppNavigationCaption :title="t('forms', 'Shared with you')" />
+				<AppNavigationForm v-for="form in sharedForms"
+					:key="form.id"
+					:form="form"
+					:read-only="true"
+					@mobile-close-navigation="mobileCloseNavigation" />
```
Comment (skjnldsv) [inline src/Forms.vue:45]: Maybe also hide if there is no "shared forms" ?

Comment (jotoeri) [reply src/Forms.vue]: 👍

**Comment by jotoeri** [comment]
Just added and already squashed one more small fix, that i just found: https://github.com/nextcloud/forms/pull/763/commits/171d43b08f5bc88d891c135358a85cd27eef64bd

File: src/Forms.vue [hunk 5]
```diff
@@ -107,26 +120,58 @@ export default {
...
+				console.error('Form not found for hash: ', this.routeHash)
+				return false
+			}
+
+			// For Routes edit or results, the form must be in owned forms.
+			if (this.$route.name === 'edit' || this.$route.name === 'results') {
+				if (this.forms.findIndex(form => form.hash === this.routeHash) < 0) {
+					return false
+				}
+			}
```
[the remark this answers is not in the document]
Comment (jotoeri) [reply src/Forms.vue:155]: Opened an issue to keep track:  #806

File: src/Forms.vue [hunk 6]
```diff
@@ -107,26 +120,58 @@ export default {
...
 		}
 	},
 
 	computed: {
 		noForms() {
-			return this.forms && this.forms.length === 0
+			return this.forms && (this.forms.length === 0) && (this.sharedForms.length === 0)
+		},
+		noOwnedForms() {
+			return this.forms && (this.forms.length === 0)
```
[the remark this answers is not in the document]
Comment (jotoeri) [reply src/Forms.vue]: First line here would not work i think, but included the rest: https://github.com/nextcloud/forms/pull/763/commits/e8ec9dc5c8a34c07a66627c3dc117fa272db3b0a

File: src/Forms.vue [hunk 7]
```diff
@@ -151,15 +196,26 @@ export default {
...
 			} catch (error) {
 				showError(t('forms', 'An error occurred while loading the forms list'))
 				console.error(error)
-			} finally {
-				this.loading = false
 			}
+
+			// Load shared forms
+			try {
+				const response = await axios.get(generateOcsUrl('apps/forms/api/v1', 2) + 'forms/shared')
```
[the remark this answers is not in the document]
Comment (jotoeri) [reply src/Forms.vue]: Yeah, i remember struggling on this, too. I changed it. 👍

File: src/Forms.vue [hunk 8]
```diff
@@ -107,26 +120,61 @@ export default {
...
 			sidebarOpened: false,
 			forms: [],
+			sharedForms: [],
 		}
 	},
 
 	computed: {
 		noForms() {
-			return this.forms && this.forms.length === 0
+			return this.noOwnedForms && this.noSharedForms
```
Comment (skjnldsv) [inline src/Forms.vue:129]: Ah right, double negative! :grin:

**Review by skjnldsv — APPROVED** [review]
Good for me! Please rebase :rocket:

**Comment by skjnldsv** [comment]
Ah, sorry, I thought I saw a fixup! ALl good then! :shipit:

File: appinfo/routes.php (hunk 1 above)
Comment (jotoeri) [reply appinfo/routes.php:32]: @rullzer, @nickvergessen, @ChristophWurst does one of you know about any problems with such generic route?
We have a few users now reporting they can not open the App via Navigation. But seems a bit specific to nginx-users 🤔

Should we eventually re-include the line without any parameters again?

Comment (nickvergessen) [reply appinfo/routes.php:32]: well I guess the route `index.php/apps/forms//` might not evaluate the same with all webservers and therefor result in problems later. I would suggest to just add multiple routes instead.

**Pull request merged**